### ISSUE TITLE
feature request: softmax target axes / multi-dimensional softmax

### ISSUE THREAD

**Issue by bmcfee** [tensorflow/tensorflow#210]
The current softmax implementation operates only on 2d tensors, and computes the normalization over the second dimension.

It would be useful to have a softmax operation that can sum over arbitrary dimensions in the same way that the `reduce_*` operations do.

As a concrete example application, one might have a convolutional network that outputs a (soft) 1-of-k prediction for each position.

**Comment by vrv** [comment]
Thanks for the request!  For the purposes of prototyping, you could probably implement this using more primitive operations (e.g., using tf.exp, tf.reduce_sum , etc) -- let us know if there's something missing from the set of primitive operations.

**Comment by bmcfee** [comment]
> Thanks for the request! For the purposes of prototyping, you could probably implement this using more primitive operations (e.g., using tf.exp, tf.reduce_sum , etc) -- let us know if there's something missing from the set of primitive operations.

Thanks!  And yeah, I've already done that for now.  It's slow and not numerically stable, but it works for now.

**Comment by georgedahl** [comment]
You should have enough ops to make it numerically stable as well. Just subtract off the result of tf.reduce_max before exponentiating.

**Comment by girving** [comment]
I think this is best implemented with `reshape` and `transpose` followed by the existing functionality.

**Comment by raingo** [comment]
I just had this need, and here is mine: https://gist.github.com/raingo/a5808fe356b8da031837
